### ISSUE TITLE
Properly model NUSIZ during player decode and draw

### ISSUE THREAD

**Issue by DirtyHairy** [stella-emu/stella#63]
...as described on [AtariAge](http://atariage.com/forums/topic/209224-requesting-help-in-improving-tia-emulation-in-stella). As for the players, this will basically boil down to replacing the precomputed player patterns with a proper description of the player scan counters. Looking at Joe Musashi's testcase, I think that there is no delay in NUSIZ (I'll have to recheck, but if I remember correctly, a delay will wreak havok on several ROMs, including Tominv5), but instead the current pixel will finish rendering and, after that, the clock divisor will continue counting at the new ratio.

**Comment by DirtyHairy** [comment]
By the way, that's a pretty fascinating thread :smirk: Several ideas from there are actually implemented in 6502.ts and the new core. A nice case of parallel evolution :beers:

**Comment by thrust26** [comment]
Sounds interesting.  Which ideas?

**Comment by DirtyHairy** [comment]
Mainly splitting the individual aspects of the TIA into separate classes and modelling the TIA as interconnected counters. I also stumbled over your branch in Stella when I first started integrating the new core (I didn't know about the corresponding thread); that stuff looks pretty similar to how the new core is structured, too.

**Comment by sa666666** [comment]
I had partly forgotten about that thread.  Looking at it now, it did lay out the framework of where we are now, but (as you say) in a completely parallel evolution.  Things were looking up at the point, but you'll notice how the thread just stopped.  In the last message, I went to PM to help the guy set up a development environment.  After a week or so of working on it, he decided that he was no longer interested.  And some of the other help that was anticipated didn't come, so I became disillusioned and (again) moved on to other parts of the codebase.

So @DirtyHairy, you'll understand why I switched to your core so quickly, as we talked about in a PM.  I've since learned that once you find reliable help, you latch on immediately :)

**Comment by thrust26** [comment]
If the same solution is found multiple times independently, that 's a very good sign.

Sorry for aborting my work on Stella, first SCA had to be done and took much longer than I had expected. And afterwards I simply was not motivated  enough to return. But at least I was able to help to forward Christian to Stella.

**Comment by sa666666** [comment]
For the record, Thomas, I wasn't referring to you in that last comment.

**Comment by DirtyHairy** [comment]
@sa666666 Indeed, this is an interesting bit of history that explains very well why you wanted to get rid of the old core as soon as possible. It is also a peculiar coincidence that the 6502.ts core came around and to fill that particular hole.

I started 6502.ts as a mere research project and tech demo on Typescript and Javascript performance more than two years ago, and after completing CPU emulation, the project went to sleep again in early 2015. It is a mere coincidence that I started work again on 2600 emulation this spring and, after getting the first few ROMs to work, I was kind of sucked in --- I wasn't into retrocomputing in general or into the VCS on particular before. The 2600 and the TIA are fascinating pieces of technology, and the more I learn about them, the more I continue to be surprised by what talented people can get out of this seemingly primitive and unwieldly hardware.

Anyway, merry christmas to you both, I'll propably be mostly gone from the web for a few days now :smirk: :christmas_tree: :snowflake:

**Comment by DirtyHairy** [comment]
Implemented in 1a7ac60 . The corresponding patterns now look like this in stella:

8 -> 16
![player8](https://cloud.githubusercontent.com/assets/2444224/22630958/49e7e3b2-ec04-11e6-8796-357cdf4e0f22.png)

8 -> 32
![player8_1](https://cloud.githubusercontent.com/assets/2444224/22630959/52b19b64-ec04-11e6-8408-cd6eb5db64d0.png)

16 -> 8
![player16](https://cloud.githubusercontent.com/assets/2444224/22630960/5a4615d0-ec04-11e6-87f7-0f0aca6df327.png)

16 -> 32
![player16_1](https://cloud.githubusercontent.com/assets/2444224/22630962/61283f2c-ec04-11e6-8cb2-2780891f1744.png)

32 -> 8
![player32](https://cloud.githubusercontent.com/assets/2444224/22630964/668f1a9e-ec04-11e6-8855-32143f9fa21a.png)

32 -> 16
![player32_1](https://cloud.githubusercontent.com/assets/2444224/22630966/6e08f5d8-ec04-11e6-9906-4a600f21a6fc.png)

**Comment by sa666666** [comment]
Just wanted to confirm that this fixes the final remaining issue in https://github.com/stella-emu/stella/issues/68 too (the second-last horizontal bar was too short).

**Comment by DirtyHairy** [comment]
This als renders Joe Musashi's [testcase](http://atariage.com/forums/topic/209224-requesting-help-in-improving-tia-emulation-in-stella/page-3#entry2703410) pixel correct.

**Comment by DirtyHairy** [comment]
However, @thrust26 's testcase is close, but there are still two deviations:

![img_20170206_224210](https://cloud.githubusercontent.com/assets/2444224/22667839/c224eb76-ecbe-11e6-8059-b5240d47dcc3.jpg)

vs.

![testsize2copies_a](https://cloud.githubusercontent.com/assets/2444224/22667869/dc6d1a08-ecbe-11e6-8a0d-2b4312b3e01b.png)

**Comment by DirtyHairy** [comment]
It seems that switching off a copy during the very early stages of decoding will skip it alltogether. fd95231 accounts for this behavior and reproduces @thrust26 's testcase perfectly.

![testsize2copies_a](https://cloud.githubusercontent.com/assets/2444224/22715520/880634bc-ed91-11e6-8cc1-03709bfa25e5.png)

Still, this has made me curious about some other cases I want to test before closing this:

* Switching from double / quad player to two close copies, inducing an overlap between between the two copies
* Systematically switching off copies just after hitting the decode
* Systematically switching on copies just before hitting the decode

**Comment by DirtyHairy** [comment]
The TIA is one hellish chip. It seems that everything that affects the counters obeys slightly different rules during hblank. I have created a new testcase that is very similar to the testcases above but executes the NUSIZ during hblank (I only 32 pixels -> x) for now: [testcase.zip](https://github.com/stella-emu/stella/files/826097/testcase.zip). As before, you can find the source in the 6502.ts repo.

The results on real hardware look like this:

32 -> 8:
![img_20170307_225903](https://cloud.githubusercontent.com/assets/2444224/23682182/cc64fa8c-0392-11e7-9436-974cf154ad98.jpg)

32 -> 16
![img_20170307_225925](https://cloud.githubusercontent.com/assets/2444224/23682186/d600cc1a-0392-11e7-898e-7e8d7de158b3.jpg)

Commit e021660 improves emulation to include this class of edga cases. The emulation now looks like this:

32 -> 8
![player32_hblank](https://cloud.githubusercontent.com/assets/2444224/23682222/0515f3a4-0393-11e7-9350-8f297c28054c.png)

32 -> 16
![player32_hblank_1](https://cloud.githubusercontent.com/assets/2444224/23682233/0efa8420-0393-11e7-98b1-60b0e3c240b7.png)

This fixes #82 for good. More to come...

**Comment by DirtyHairy** [comment]
The very same commit also fixes the cases 16 -> 8 and 16 -> 32. For reference:

16 -> 8 hardware
![16_to_8](https://cloud.githubusercontent.com/assets/2444224/23786719/ddc75576-056e-11e7-999c-97edba24acc7.jpg)

16 -> 32 hardware
![16_to_32](https://cloud.githubusercontent.com/assets/2444224/23786737/eb6d28c2-056e-11e7-95c9-b22bf8d06a5c.jpg)


16 -> 8 Stella
![player16_hblank](https://cloud.githubusercontent.com/assets/2444224/23786767/1923c56e-056f-11e7-9023-8091903a2e6c.png)

16 -> 32 Stella
![player16_hblank_1](https://cloud.githubusercontent.com/assets/2444224/23786780/286ee53a-056f-11e7-9c99-025fbfe99251.png)

**Comment by thrust26** [comment]
@DirtyHairy Is this still an issue?

**Comment by DirtyHairy** [comment]
Not so much an issue as an reminder that there are still some edge cases that I didn't investigate.
